.github/workflows/validate.yml

```diff
@@ -426,12 +426,10 @@ jobs:
           name: cabal-macOS-x86_64
 
       - name: Create GitHub prerelease
-        uses: marvinpinto/action-automatic-releases@v1.2.1
+        uses: softprops/action-gh-release@v2
         with:
-          repo_token: ${{ secrets.GITHUB_TOKEN }}
```

Comment:
I think I'm going to be surprised if this doesn't require the token to be passed to it: actions are restricted as to what they're allowed to do to repos.

Reply:
As I said: there's a default value that looked like it might work okay. But I don't have a good idea how these work, so I may be completely off.

Reply:
https://docs.github.com/en/actions/sharing-automations/avoiding-duplication note the bottom of the second column in the table comparing reusable workflows to composite actions. _You must explicitly enable use of workflow secrets by an action._

Reply:
https://github.com/softprops/action-gh-release/tree/v2/?tab=readme-ov-file#permissions

Reply:
I still don't understand why the default value is not good but I have put back the token property.

Reply:
I'm sorry, what part of "tokens are not available to actions" is unclear? (Exercise: consider reasons why this might be the case. Remember that actions are code not under your control.)

Reply:
No, you're good. If the token isn't available to actions, how did the previous action manage?

Reply:
The line you removed that explicitly passed the token to the action. (As per the documentation I linked above, the new one doesn't work that way; it uses a different mechanism to gain permissions.)

Reply:
@Kleidukos' workflow uses this action without explicit token management: https://github.com/Kleidukos/print-api/blob/main/.github/workflows/ci.yml Also, no permissions are set in the workflow file. Could we not do the same?

Reply:
The first link you gave seems beside the point as we don't use Reusable workflows or Composite actions, to the best of my knowledge.

Reply:
This is _exactly_ a composite action: you invoke it with `uses:`.

Reply:
Great! How can we move forward?

Reply:
TBH I'm not sure. Looking at `action.yml` it apparently _can_ be passed a token, but has a default; but GHA's documentation for (composite) actions says it shouldn't have a default. There's a later line that again assumes it receives a token by default. This might be a question for GitHub Support.

(I should note that GitHub's official documentation calls them "composite actions" to distinguish from other uses of "actions", specifically including the GitHub Actions service. But if you compare [Creating a composite action](https://docs.github.com/en/actions/sharing-automations/creating-actions/creating-a-composite-action) to actual actions (for example, `softprops/action-gh-release@v2` is https://github.com/softprops/action-gh-release/tree/v2/), they're pretty obviously the same thing.)

That leaves the claim in https://github.com/softprops/action-gh-release/tree/v2/?tab=readme-ov-file#permissions that the token has to be given write permissions. I've already confirmed the standard token doesn't have that permission by default: compare https://github.com/haskell/cabal/pull/10271 to https://github.com/ubuntu/authd/commit/3f9df8f21d952cd33fd44d3834d0edeec1f5766f, where I had to drop `marocchino/sticky-pull-request-comment@v2` because it requires `permissions: write` and I was unable to figure out how to actually grant that permission to `GITHUB_TOKEN`. (I concluded we didn't really need a sticky comment; the job's status was good enough, and didn't track down how granting permissions to tokens actually works as a result.)

Reply:
Did you see my earlier comment about the `print-api` repository? It doesn't pass any token to the action and works fine, apparently.

Reply:
Yes, see

> can be passed a token, but has a default; but GHA's documentation for (composite) actions says it shouldn't have a default

The current question is what permissions the token it is apparently being passed anyway has.

Reply:
I should note that, if the default token _does_ in fact have permissions to make releases, what other permissions does it have that might be problematic with respect to random composite actions? Does someone need to audit the actions we use regularly to make sure they're safe? The actual (as opposed to documented) permissions structure seems to lean a bit too far toward convenience over security.

Reply:
(I want to leave this unresolved as a reminder to contact GitHub support, but with respect to this PR specifically it's resolved at this point.)

Reply:
It took me an hour to figure out a configuration that works using this testing repository: https://github.com/ulysses4ever/test-prerelease tldr; `permissions:contents:write` seems enough.

Would you be interested in re-reviewing?

Reply:
I may be able to get to it later today; I just spent my available time on the changelogs review, and have other things I need to get done now.